src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,153 @@
+4. Get the value in `data` column whose timestamp is exactly `7`. We got `$3`.
+
+This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will see the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.
+
+### Handling Conflicts
+
+Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.
+
+When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.
+
+To roll back a row, just simply remove its lock and its corresponding value in `data` column.
+
+### Tolerating crashes
+
+Percolator has the ability to survive crashes without breaking the correctness of data.
+
+First, let's see what will happen after a crash. A crash may happen during `Prewrite`, during `Commit` or between these two. But anyway, we can simply divide these conditions into two types: before committing the primary, or after committing the primary.
+
+So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it, in fact, doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.
+
+* If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.
+* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
+* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.
+* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
+
+## Percolator in TiKV
+
+TiKV is a key-value storage, and each key-value pair can be regarded as a row in Percolator.
+
+TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.
```

Comment:
```suggestion
TiKV internally uses RocksDB, a stand-alone key-value storage engine library, to persist data to local disk. RocksDB's atomic write batch and TiKV's transaction scheduler jointly make it atomic to read and write a single user key, which is a requirement of Percolator.
```

Reply:
RocksDB is a library but not a stand-alone process.
"atomic" is a term to say an operation is "non-dividable".

Reply:
Got it. Then the original sentence is fine.